Refactor dada

bgpalmer opened the pull request.

m-murphy commented on bin/dada2_denoising.R:
```diff
@@ -0,0 +1,136 @@
+save(dadaRs, file="dada_denoised_R.RData")
+
+# Filter samples based on the provided condition
+filter_samples <- function(samples, condition) {
```
small nit -- I prefer putting all function declarations at the top of the script, after the argument parsing

bgpalmer replied:
- [ ] functions at top

m-murphy commented on bin/dada2_denoising.R:
```diff
@@ -0,0 +1,136 @@
+  mergers <- merge_with_concatenation(dadaFs, derepFs, dadaRs, derepRs, amplicon.info)
+  
+  sample_name <- get_sample_name(names(derepFs)[1], loci_list)  # extract sample name
+  saveRDS(mergers, file = paste0(sample_name, "_merged.RDS"))
```
I think we should add an output dir option to the script

bgpalmer replied:
- [ ] add dout

m-murphy commented on bin/dada2_dereplicate.R:
lgtm

m-murphy commented on bin/dada2_postprocessing.R:
```diff
@@ -0,0 +1,79 @@
+parser$add_argument('--seqtab-paths', nargs='+', type="character", required=TRUE, help="Paths to RDS files containing DADA2 sequence tables")
+parser$add_argument('--amplicon-file', type="character", required=TRUE, help="Path to the amplicon file")
+parser$add_argument('--ncores', type="numeric", default=1, help="Number of threads to use")
+parser$add_argument('--verbose', action='store_true', help="Add verbosity")
```
let's add a --dout arg

bgpalmer replied:
done

m-murphy reviewed (commented):
Everything is looking good I think.  One thing overall I think we should consider is our logging approach. With computing on AWS, print statements go into the logging systems, so it's a bit more important to take care with how/when we print things. I think a logging library could be useful here to set different levels and control what gets printed out for interactive use vs on cloud compute https://daroczig.github.io/logger/

bgpalmer commented on bin/dada_overlaps.R:
- [ ] yes this should be removed

bgpalmer commented on bin/dada2_learn_errors.R:
```diff
@@ -0,0 +1,34 @@
+}
+
+# Learn errors for 1 or Merged using all loaded files
+error_model <- learnErrors(args$filt_1, multithread=args$ncores, MAX_CONSIST=args$maxConsist, randomize=args$rand)
```
- [ ] add argument

bgpalmer replied:
I'll add this is as a separate issue to keep track of it

bgpalmer commented on bin/dada2_preprocessing.R:
i don't have all of the steps benchmarked but I would say it takes a moderate amount of time. it sounds like phix removal would be the critical path to focus on, i'll see if there is some room for improvement there

- [ ] check phix removal in dada

bgpalmer commented on bin/dada2_denoising.R:
```diff
@@ -0,0 +1,136 @@
+library(argparse)
+
+parser <- ArgumentParser(description='DADA2 Denoising')
+parser$add_argument('--derep-1', type="character", required=TRUE, nargs="+", help="Path to RDS file containing dereplicated forward sequences")
```
this is all of the derep files for a sample - there will be a derep per amplicon. i'll update the help

- [ ] update help

bgpalmer commented on bin/dada2_preprocessing.R:
```diff
@@ -0,0 +1,72 @@
+filtRs_list <- mapply(function(op, sn) paste0(op, "/", sn, "_R_filt.fastq.gz"), output_paths, sample_names_list, SIMPLIFY = FALSE)
+
+# Perform Filtering and save filter metadata
+filter_metadata_list <- mapply(function(fnFs, fnRs, filtFs, filtRs) {
```
not really a memory thing, i thought it may be useful to write this in a way so the script could accept data from multiple samples or a single sample, depending on the use case.  i don't really do that in the other scripts, i just assume it's data for a single sample, although they could be modified so that data for multiple samples could be passed in

m-murphy replied:
Oh I see, so this pattern makes it so it can run on one file or a directory? If it doesn't cause a perf hit that makes sense to me

bgpalmer replied:
mostly just more than one directory. i thought writing it this way would allow some flexibility. there shouldn't be any added performance problems from this, just an expectation that runtime will grow linearly if you decide to pass in more than one sample directory

bgpalmer commented:
@m-murphy - I believe I've addressed all of the comments. I added in logging using the logging module you found (good find btw). Let me know if you have othere concerns

m-murphy reviewed (approved):
Looks great! Now that we have the logging in place, what do you think about adding in timing info too? that would make it easier to identify where things take a long time

bgpalmer commented:
@m-murphy - added logging to track function memory usage and duration, particularly for dada scripts. For memory usage, I went with the lobstr package

m-murphy commented:
> @m-murphy - added logging to track function memory usage and duration, particularly for dada scripts. For memory usage, I went with the lobstr package

ooh fun
